perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -1922,4 +1969,188 @@ private static void prepareServiceRoles(PerunSession sess) {
+		//Traverse through outer role list which works like logical OR
+		for (JsonNode roleArray: policyRoles) {
+
+			boolean authorized = true;
+			//Traverse through inner role list which works like logical AND
+			Iterator<String> roleArrayKeys = roleArray.fieldNames();
+			while (roleArrayKeys.hasNext()) {
+				String role = roleArrayKeys.next();
+
+				//fetch the role, object is connected with, from config
+				JsonNode roleObject = roleArray.get(role);
+
+				// If policy role is not connected to any object or if there is no corresponding type of object in the perunBeans map
+				if (roleObject.isNull() || !mapOfBeans.containsKey(roleObject.textValue())) {
+					//If principal does not have the role, this inner list's result is false
+					if (!sess.getPerunPrincipal().getRoles().hasRole(role)) {
+						authorized = false;
+						break;
+					}
+					// If policy role is connected to some object, like VOADMIN->Vo
+				} else {
+					//traverse all related objects from perun which are relevant for the authorized method
+					for (Integer objectId : mapOfBeans.get(roleObject.textValue())) {
+						//If the principal does not have rights on role-object, this inner list's result is false
+						if (!sess.getPerunPrincipal().getRoles().hasRole(role, roleObject.textValue(), objectId)) {
+							authorized = false;
+							break;
+						}
+					}
+					//Some inner role check failed so jump out of the loop
+					if (!authorized) break;
+				}
+			}
+			// If all checks for inner role list pass, return true. Otherwise proceed to another inner role list
+			if (authorized) return true;
+		}
+		//If all checks failed return false
+		return false;
+	}
+
+	/**
+	 * Fetch all possible PerunBeans for each of the objects from the list according to the id of the bean in the object.
+	 *
+	 * @param objects for which will be related objects fetched.
+	 * @return all related objects together with the objects from the input as a map of PerunBean names and ids.
+	 */
+	private static Map<String, List<Integer>> fetchAllRelatedObjects(List<PerunBean> objects) {
+		List<PerunBean> relatedObjects = new ArrayList<>();
+		//Create a map from objects for easier manipulation and duplicity prevention
+		Map<String, List<Integer>> mapOfBeans = new HashMap<>();
+
+		for (PerunBean object: objects) {
+			relatedObjects.add(object);
+			List<PerunBean> retrievedObjects = RelatedObjectsResolver.getValue(object.getBeanName()).apply(object);
+			relatedObjects.addAll(retrievedObjects);
+		}
+
+		//Fill map with PerunBean names as keys and a lists of unique ids as value for each bean name
+		for (PerunBean object : relatedObjects) {
+			if (!mapOfBeans.containsKey(object.getBeanName())) {
+				List<Integer> listOfBeanIds = new ArrayList<>();
+				listOfBeanIds.add(object.getId());
+				mapOfBeans.put(object.getBeanName(), listOfBeanIds);
+			} else if (!mapOfBeans.get(object.getBeanName()).contains(object.getId()))
+				mapOfBeans.get(object.getBeanName()).add(object.getId());
```

Comment:
NIT: In the case of "HashSet" instead of "ArrayList", you can skip the part of checking and in the else part add the bean id always. But this is just a tip, you don't need to change that if you don't want to. It is still working, just the "else if" branch is harder to read then it could be.

Reply:
Changed to HashSet